Modules/GLO/src/ITSTPCMatchingTask.cxx

```diff
@@ -151,45 +151,57 @@ void ITSTPCMatchingTask::endOfCycle()
+      std::string name = eff->GetName();
+      name += "_Hist";
+      auto ratio = std::make_unique<common::TH1FRatio>(name.c_str(), eff->GetTitle(),
+                                                       eff->GetPassedHistogram()->GetXaxis()->GetNbins(),
+                                                       eff->GetPassedHistogram()->GetXaxis()->GetXmin(),
+                                                       eff->GetPassedHistogram()->GetXaxis()->GetXmax());
+      ratio->SetBit(TH1::EStatusBits::kNoStats);
+      if (!ratio->getNum()->Add(eff->GetPassedHistogram())) {
+        ILOG(Error) << "Add operation for numerator histogram of " << name << " failed; efficiency will be skewed" << ENDM;
       }
-    }
-
-    { // Eta
-      auto hEffEta = mMatchITSTPCQC.getFractionITSTPCmatchEta(gloqc::MatchITSTPCQC::ITS);
-      mHEffEta.reset();
-      mHEffEta.reset(dynamic_cast<TH1*>(hEffEta->GetPassedHistogram()->Clone("mFractionITSTPCmatchEta_ITS_Hist")));
-      if (mHEffEta) {
-        mHEffEta->Divide(hEffEta->GetPassedHistogram(), hEffEta->GetTotalHistogram(), 1.0, 1.0, "B");
-        mHEffEta->SetBit(TH1::EStatusBits::kNoStats);
-        getObjectsManager()->startPublishing(mHEffEta.get(), PublicationPolicy::Once);
-      } else {
-        ILOG(Error) << "Failed cast for hEffEtaHist, will not publish!" << ENDM;
+      if (!ratio->getDen()->Add(eff->GetTotalHistogram())) {
+        ILOG(Error) << "Add operation for denominator histogram of " << name << " failed; efficiency will be skewed" << ENDM;
       }
-    }
-
-    { // Phi
-      auto hEffPhi = mMatchITSTPCQC.getFractionITSTPCmatchPhi(gloqc::MatchITSTPCQC::ITS);
-      mHEffPhi.reset();
-      mHEffPhi.reset(dynamic_cast<TH1*>(hEffPhi->GetPassedHistogram()->Clone("mFractionITSTPCmatchPhi_ITS_Hist")));
-      if (mHEffPhi) {
-        mHEffPhi->Divide(hEffPhi->GetPassedHistogram(), hEffPhi->GetTotalHistogram(), 1.0, 1.0, "B");
-        mHEffPhi->SetBit(TH1::EStatusBits::kNoStats);
-        getObjectsManager()->startPublishing(mHEffPhi.get(), PublicationPolicy::Once);
-      } else {
-        ILOG(Error) << "Failed cast for hEffPhiHist, will not publish!" << ENDM;
+      ratio->Sumw2();
+      ratio->update();
+
+      // Calculate binominal errors for efficiency, taken from ROOT
+      for (Int_t iBin = 0; iBin < ratio->GetNcells(); ++iBin) {
+        const Double_t b1 = ratio->getNum()->GetBinContent(iBin);
+        const Double_t b2 = ratio->getDen()->GetBinContent(iBin);
+        if (b2 == 0) {
+          ratio->SetBinError(iBin, 0);
+          continue;
+        }
+        const Double_t b1sq = b1 * b1;
+        const Double_t b2sq = b2 * b2;
+        const Double_t e1 = ratio->getNum()->GetBinError(iBin), e1sq = e1 * e1;
+        const Double_t e2 = ratio->getDen()->GetBinError(iBin), e2sq = e2 * e2;
+        if (b1 != b2) {
+          ratio->SetBinError(iBin, TMath::Sqrt(TMath::Abs(((1. - 2. * b1 / b2) * e1sq + b1sq * e2sq / b2sq) / b2sq)));
```

Comment:
Do you know what happens to the errors when two objects are merged?

Reply:
Ah your right, when merged this would be done incorrectly... This should probably be done in the TH1Ratio::update under some flag. Which I guess I can do but do not know if I should?

Reply:
I think it's fine, but please do it in a separate PR and I will also ask Andrea to have a look.

Reply:
sure, dropped it for now.